xtm/xtm_api.h

```diff
@@ -0,0 +1,159 @@
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <stdlib.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/** \cond public */
+
+/**
+ * Opaque struct, that represent one-directional, one-reader-one-writer
+ * queue implementation with eventloop integration ability.
+ */
+struct xtm_queue;
+
+/**
+ * Allocation function, used to create instance of struct xtm_queue.
+ * @param[in] size - queue size, must be power of two
+ * @retval returns pointer to new xtm_queue or NULL in case of error,
+ *         with the following possible errno: ENOMEM, EINVAL, EMFILE,
+ *         ENFILE, ENODEV.
+ */
+struct xtm_queue *
+xtm_create(unsigned size);
+
+/**
+ * Deallocation function, used to free queue and close its internal fds.
+ * @param[in] queue - xtm_queue to delete
+ * @retval returns 0 on success. Otherwise -1 with errno set appropriately
+ *         (as in close(2), since it implies close of fds).
+ */
+int
+xtm_delete(struct xtm_queue *queue);
+
+/**
+ * Function for queue consumer notification.
+ * @param[in] queue - xtm_queue to notify
+ * @retval returns 0 on success. Otherwise -1 with errno set appropriately
+ *         (as in write(2), since it implies write to internal fd).
+ */
+int
+xtm_msg_notify(struct xtm_queue *queue);
+
+/**
+ * Function checks is there are free space in queue
+ * @param[in] queue - xtm_queue to ckeck free space
+ * @retval returns 0 if queue has space. Otherwise -1 with errno set
+ *         to ENOBUFS.
+ */
+int
+xtm_msg_probe(struct xtm_queue *queue);
+
+/**
+ * Function returns current count of messages in xtm queue
+ * @param[in] queue - xtm_queue to check current count of messages
+ * @retval returns count of messages in queue.
+ */
+unsigned
+xtm_msg_count(struct xtm_queue *queue);
+
+/**
+ * Function puts a message containing the function and
+ * its argument to the queue
+ * @param[in] queue - xtm_queue to put message
+ * @param[in] fun - message function
+ * @param[in] fun_arg - message function argument
+ * @param[in] delayed - flag indicating the consumer should be notified
```

Comment:
For flags we always use 'bool' and the name must start with 'is_' prefix. But as I said, I don't like that flag regardless of the name. When you will call that function like this: `xtm_fun_dispatch (queue, fun, arg, 0)`, it is very unclear what is '0', and it adds branching inside even if you know the flag at compile time.